### PR TITLE
Replace the deprecated IndexReader APIs with new storedFields() & termVectors()

### PR DESCRIPTION
**Pull request by luyuncheng**
i try to use OpenSearch in production. and in our environment, we have tons of segments, tons of threads, tons of data. 

i see there is deprecated API to retrieve document. These API would create many `ThreadLocal StoreFields` and make jvm can not GC. you can see the example in LUCENE#11987. 

so i PR the code to replace the deprecated api with new `storefields` & `termVectors ` API and reduced threadlocal creation.

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/TranslogLeafReader.java [hunk 1]
```diff
@@ -221,28 +221,33 @@ public int maxDoc() {
 
     @Override
     public void document(int docID, StoredFieldVisitor visitor) throws IOException {
-        if (docID != 0) {
```
Comment (gashutos) [inline server/src/main/java/org/opensearch/index/engine/TranslogLeafReader.java:224]: Lets this implemenation be same as is and only add `storeField` implementation ?

**Review by gashutos — APPROVED** [review]
Thanks @luyuncheng just a minor comment but looks good to me !
There are no other places we use these 2 deprecated APIs righ ?

Also tagging @dblock  / @reta for additional reviews.

File: server/src/main/java/org/opensearch/index/engine/TranslogLeafReader.java (hunk 1 above)
Comment (luyuncheng) [reply server/src/main/java/org/opensearch/index/engine/TranslogLeafReader.java:224]: As `TranslogLeafReader` has it own storedFields/document implementation,  so i put the logic into `storedFields`. and implement  `StoredFields#document`  which is new lucene interface required  see the following part
https://github.com/opensearch-project/OpenSearch/blob/c18cd7fb34f7a37c56ce26608073ec75f9600023/server/src/main/java/org/opensearch/index/engine/TranslogLeafReader.java#L229-L231

**Comment by gashutos** [comment]
> >  Thanks @luyuncheng just a minor comment but looks good to me !
> 
> There are no other places we use these 2 deprecated APIs right ?
> 
> 
> 
> @gashutos i think there is no deprecated APIs calling. we deployed a environment and then dump heap with write/search reqeusts, there is no threadlocal stroefields. 
> 
> 

Got it.

**Comment by luyuncheng** [comment]
Let me know if there is anything I can do to support this? Our team made these change into production environment and supported large number of shards. @dblock  @reta  @gashutos

**Comment by luyuncheng** [comment]
> I think the description of the change is a bit misleading, may be change it to:

> We also need a CHANGELOG entry please (see please CHANGELOG.md), thank you

++ Updated
@reta  thanks for the review

**Comment by luyuncheng** [comment]
> conflicts :(, needs rebase with main

@reta thanks. i sync upstream master into this branch, fixed conflicts. and run precommit locally seems success.

**Comment by reta** [comment]
@luyuncheng sadly automation failed, could you please send manual backport to `2.x` branch? thank you

**Comment by luyuncheng** [comment]
> could you please send manual backport to 2.x branch? thank you
@reta  OK

**Pull request merged**